wire/msgtx.go

```diff
@@ -784,37 +997,57 @@ func (msg *MsgTx) BtcDecode(r io.Reader, pver uint32) error {
 	msg.Version = int32(version)
 	_, mType := msgTxVersionToVars(msg.Version)
 
+	// returnScriptBuffers is a closure that returns any script buffers that
+	// were borrowed from the pool when there are any deserialization
+	// errors.  This is only valid to call before the final step which
+	// replaces the scripts with the location in a contiguous buffer and
+	// returns them.
+	returnScriptBuffers := func() {
+		for _, txIn := range msg.TxIn {
+			if txIn == nil || txIn.SignatureScript == nil {
+				continue
+			}
+			scriptPool.Return(txIn.SignatureScript)
+		}
+		for _, txOut := range msg.TxOut {
+			if txOut == nil || txOut.PkScript == nil {
+				continue
+			}
+			scriptPool.Return(txOut.PkScript)
+		}
+	}
+
 	switch {
 	case mType == TxSerializeNoWitness:
-		err := msg.decodePrefix(r, pver)
+		err := msg.decodePrefix(r, pver, returnScriptBuffers)
```

Comment:
Rather than passing the `returnScriptBuffers` func to all of these and calling it in their error paths, I propose just calling it from this function in the `if err != nil {` blocks.  That way the called funcs can't forget to invoke it since they aren't responsible for it.